utils/vector.hpp

```diff
@@ -0,0 +1,366 @@
+#pragma once
+
+namespace nstd
+{
+template<class T>
+class Vector
+{
+public:
+    using uint = unsigned int;
+    using reference = T&;
+    using iterator  = T*;
+    using const_reference = const T&;
+    using const_iterator  = const T*;
+
+
+    /* Constructors */
+    Vector();
+    Vector(const Vector<T>&);
+    Vector(Vector<T>&&);
+    explicit Vector(uint);
+
+
+    /* Destructor */
+    ~Vector();
+
+
+    /* Capacity */
+    uint getSize() const;
+    uint getCapacity() const;
+    void resize(uint);
+    void resize(uint, const_reference);
+    void reserve(uint);
+
+
+    /* Member Functions */
+    void push_back(const_reference);
+    void pop_back();
+    void clear();
+
+
+    /* Operator Overloads */
+    Vector<T>& operator=(const Vector&);
+    Vector<T>& operator=(Vector&&);
+    reference  operator[](uint);
+
+
+    /* Iterators */
+    iterator begin();
+    iterator end();
+    const_iterator begin() const;
+    const_iterator end()   const;
+    iterator erase(iterator);
+    iterator insert(const_iterator, const_reference);
+
+
+
+private:
+    /* Helpers */
+    bool iteratorInVector(iterator);
+
+
+    /* Data */
+    T* array;
+    uint size;
+    uint capacity;
+};
+
+
+/* Constructors */
+template <class T>
+Vector<T>::Vector()
+    : size(0), capacity(0), array(nullptr)
```

Comment:
Let's default capacity to some decent-sized power of 2 -- I'm thinking 16? And let's then initialize array to a dynamic array of size `size`